### PR TITLE
Prevent data overwrite by old_names in register_messages

### PR DESCRIPTION
**Pull request by glennmatthews**
### Fixes / new features
-  Add missing validation of message ``old_names`` in ``MessagesStore.register_messages()`` to prevent a newly registered message from overwriting values previously registered by another message.
- Add `InvalidMessageError` to raise in these and similar cases instead of using `assert`.
- Move `EmptyReport` and `UnknownMessage` exceptions (and `InvalidMessageError`) to `pylint.exceptions` submodule.
- Rename `EmptyReport` to `EmptyReportError` and `UnknownMessage` to `UnknownMessageError`

### CODE REVIEW

File: pylint/utils.py [hunk 1]
```diff
@@ -693,6 +693,16 @@ def register_messages(self, checker):
...
+                assert (old_id not in self._alternative_names or
+                        self._alternative_names[old_id] == msg), \
+                        'Old message id %r is already defined' % old_id
+                assert (old_symbol not in self._messages or
+                        self._messages[old_symbol] == msg), \
+                        'Old message symbol %r is already defined ' \
+                        'as a current symbol' % old_symbol
+                assert (old_symbol not in self._alternative_names or
+                        self._alternative_names[old_symbol] == msg), \
+                        'Old message symbol %r is already defined' % old_symbol
```
Comment (AWhetter) [inline pylint/utils.py]: We should raise something ourself rather than use assertions.

Comment (glennmatthews) [inline pylint/utils.py]: Fair enough. I was following the existing model (just above this code) but am happy to rework this and the above logic to use proper exceptions. Suggestions for what Exception class to use (or define)?

Comment (PCManticore) [inline pylint/utils.py]: exceptions.InvalidMessageError?

Comment (glennmatthews) [inline pylint/utils.py]: Looks like utils.py already has `UnknownMessage` and `

Comment (PCManticore) [inline pylint/utils.py]: I think we can safely move these two into a new exceptions.py module.

Comment (PCManticore) [inline pylint/utils.py]: Oh, I would prefer to end in Error, the other two could be updated as well.

Comment (glennmatthews) [inline pylint/utils.py]: That'll expand the scope of this PR a fair bit but I'll give it a shot.

Comment (PCManticore) [inline pylint/utils.py]: Oh, I was referring in general, not necessarily to do it in this PR. If you still do it though, make sure they are different commits.

Comment (glennmatthews) [inline pylint/utils.py]: OK, so:

commit 1: Create exceptions.py, add InvalidMessageError and the error checking that will raise this error.
commit 2: Move UnknownMessage and EmptyReport to exceptions.py as well
commit 3: Rename to UnknownMessageError and EmptyReportError throughout the code

Does that sound right?

Comment (PCManticore) [inline pylint/utils.py]: Yep!

**Comment by glennmatthews** [comment]
Reworked diffs as suggested. Thanks!

File: pylint/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
```
Comment (PCManticore) [inline pylint/exceptions.py]: Shouldn't we add a Copyright notice here? We just had some issues with Debian packaging due to the fact that we were missing Copyright notices on files. I am not sure what Copyright this file should have, yours or Logilab's? I think yours, no?

**Comment by PCManticore** [comment]
Looks amazing, thanks! There is only one issue left to address, with regard to a copyright notice.

File: pylint/exceptions.py (hunk 2 above)
Comment (glennmatthews) [inline pylint/exceptions.py]: No idea. :-) I only see copyright headers on a few files in the whole repo - are you sure this is needed?

Comment (PCManticore) [inline pylint/exceptions.py]: Yep, I just added them, do a git pull and it should be everywhere.

Take a look at checkers/spelling.py, I think it should be in the same line.

Comment (glennmatthews) [inline pylint/exceptions.py]: Gotcha. Added.

**Comment by AWhetter** [comment]
I'm being super picky but I'd prefer not to rename `EmptyReport` and `UnknownMessage`, and to drop `Error` from the name of `InvalidMessageError`. Because they're exceptions I already know they are an error when they are thrown. And when we use them in code it's obvious from context that they're exceptions because they'll either be `raise`d or used in an except handler. So adding `Error` on the end doesn't give a reader useful information. And typing 5 extra characters is just too much extra work for me! :wink:

Like I said, I'm being super picky. So if it's not too much trouble to change then please change it. If you don't do it then I don't mind.

**Comment by PCManticore** [comment]
Let's merge it as is. I actually prefer having Error in the names of Exceptions, not just for conveying their meaning when reading the code, but also for having them standardized and helping with introspection (when you dir() a module, what is a given class X?). To me, whenever a class doesn't have it, it suggests there might be some error and I have to check if it has Exception in mro or not.

**Comment by PCManticore** [comment]
Sorry @AWhetter :-(

**Comment by AWhetter** [comment]
That's totally fine. No need to apologise. We all have different likes and dislikes. Especially when it comes to code :smile:
